README.md
# Slog

Print multiple values as string (except functions and binaries) for debugging. Slog takes a list of values or a single value that is printed as string for debugging. Any value (string, atom, map, list, struct, keyword list ...) except functions can be printed as string with slog.

## Usage:

Slog takes a list of values or a single value that is printed as string for debugging. It takes optional keyword list as second argument

```elixir
iex> Slog.log ["value is ", {:x, :x, :x}, [key: "value"], {:ok, "Hello Universe!"}]
"value is  {:x, :x, :x} [{:key, value}] {:ok, Hello Universe!}"

# using configuration options 
iex> Slog.log ["hello", "universe"], stdout: false, delimiter: "---"
"hello---universe"

iex> Slog.log ["value is ", {:x, :x, :x}, [key: "value"], {:ok, "Hello Universe!"}, %User{age: 29, name: "Arun"}]
"value is  {:x, :x, :x} [{:key, value}] {:ok, Hello Universe!} %User{age: 29, name: Arun}"

# can take an optional delimiter to be printed in between values
iex> Slog.log ["value is ", {:x, :x, :x}, [key: "value"], {:ok, "Hello Universe!"}], delimiter: " --> "
"value is  --> {:x, :x, :x} --> [{:key, value}] --> {:ok, Hello Universe!}"

iex> Slog.log {:hello, "Universe!"}
"{:hello, Universe!}"

iex> Slog.log %{name: "Arun", age: 29}
"%{age: 29, name: Arun}"

iex> Slog.log %{manager: %User{name: "Leonardo", age: 69}}
"%{manager: %User{age: 69, name: Leonardo}}"
```

Takes a keyword list as second parameter for options. Right now, only configurable option is `:delimiter` which takes a string and prints it between each value of the list. Usage is `Slog.log [{:hello}, {:universe}], delimiter: " --> "`


## Configuration Options:

| option          | type           | default  | comments |
| ----------------|:--------------:| --------:| ---------|
| delimiter       | string         | " " (single space)        |   Default delimiter is single space       |
| stdout          | boolean        | true     |   By default `Slog.log` prints to stdout using IO.puts. Try ```Slog.log ["your_log_value"] stdout: false``` to turn off default logging to stdout       |

## Installation

[Slog](https://hexdocs.pm/slog/) can be installed
by adding `slog` to your list of dependencies in `mix.exs`:


```elixir
def deps do
  [
    {:slog, "~> 0.1.0", only: [:dev, :test]}
  ]
end
```

## About
`Slog` is an experimental module that helps printing multiple values together instead of using something like IO.inspect multiple times. For now it is recommended to use only in `:dev, :test` environments. Feedbacks and contributions appreciated.

## CHANGELOG

[View Changelog](Changelog.md)

## LICENSE

MIT

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at [https://hexdocs.pm/slog](https://hexdocs.pm/slog).



Changelog.md
## 0.1.1
Optional configuration `stdout`. Default is `true`.

## 0.1.0
First Version. Print single/multiple values as string. Configurable options - `:delimiter`

test/slog_test.exs
defmodule SlogTest do
  use ExUnit.Case
  doctest Slog

  test ":atom ==> string" do
    assert Slog.log :atom == ":atom"
  end
end


test/user_struct_test.exs
defmodule User do
  defstruct name: "Arun", age: 29
end

mix.exs
defmodule Slog.MixProject do
  use Mix.Project

  def project do
    [
      app: :slog,
      version: "0.1.1",
      description: "Print multiple values as string (except functions and binaries) for debugging.",
      elixir: "~> 1.6",
      start_permanent: Mix.env() == :prod,
      deps: deps(),

      # ExDoc steps
      # Docs
      name: "Slog",
      source_url: "https://github.com/palerdot/slog",
      homepage_url: "https://github.com/palerdot/slog",
      docs: [main: "readme", # The main page in the docs
          extras: ["README.md"]],

      # mix package stuffs
      package: [
        maintainers: ["Arun Kumar <palerdot@gmail.com>"],
        licenses: ["MIT"],
        links: %{"GitHub" => "https://github.com/palerdot/slog"}
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      # {:dep_from_hexpm, "~> 0.3.0"},
      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"},
      {:ex_doc, "~> 0.19", only: :dev, runtime: false}
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :slog, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:slog, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/slog.ex
defmodule Slog do
  @moduledoc """
  Documentation for Slog.

  Print multiple values as string (except functions and binaries) for debugging. Slog takes a list of values or a single value that is printed as string for debugging. 
  Any value (string, atom, map, list, struct, keyword list ...) except functions can be printed as string with slog.
  """
  @default_options [delimiter: " ", stdout: true]

  @doc """
  logs given list of data, takes optional keyword list

  ## Examples

      # define user struct 
      defmodule User do
        defstruct name: "Arun", age: 29
      end

      # iex> import User
      # User

      iex> Slog.log ["value is ", {:x, :x, :x}, [key: "value"], {:ok, "Hello Universe!"}]
      "value is  {:x, :x, :x} [{:key, value}] {:ok, Hello Universe!}"

      iex> Slog.log ["hello", "universe"], stdout: true, delimiter: "---"
      "hello---universe"

      iex> Slog.log ["value is ", {:x, :x, :x}, [key: "value"], {:ok, "Hello Universe!"}, %User{age: 29, name: "Arun"}]
      "value is  {:x, :x, :x} [{:key, value}] {:ok, Hello Universe!} %User{age: 29, name: Arun}"

      iex> Slog.log ["value is ", {:x, :x, :x}, [key: "value"], {:ok, "Hello Universe!"}], delimiter: " --> "
      "value is  --> {:x, :x, :x} --> [{:key, value}] --> {:ok, Hello Universe!}"

      iex> Slog.log {:hello, "Universe!"}
      "{:hello, Universe!}"

      iex> Slog.log %{name: "Arun", age: 29}
      "%{age: 29, name: Arun}"

      iex> Slog.log %{manager: %User{name: "Leonardo", age: 69}}
      "%{manager: %User{age: 69, name: Leonardo}}"
  """
  def log(data, opts) when is_list(data) do
    
    # v1 -> syncronous processing
    logged = Enum.map data, &process_elements/1

    # v2 -> async processing
    # logged = 
    #   data
    #   |> Stream.map(&async_query/1)
    #   |> Enum.map(&get_result/1)

    # construct options for logging
    options = Keyword.merge(@default_options, opts)
    delimiter = Keyword.fetch!(options, :delimiter)
    is_stdout = Keyword.fetch!(options, :stdout)

    # log the final result
    final_result = Enum.join(logged, delimiter)

    # if stdout is true log to the 'stdout'
    # SIDE EFFECT!!!
    if (is_stdout) do
      IO.puts final_result
    end

    # return the final result
    final_result
  end
  def log(data) when is_list(data) do
    log(data, @default_options)
  end
  def log(d) do
    log([d], @default_options)
  end

  # called when not a valid argument
  def log() do
    raise SlogArgumentError
  end

  # private functions

  # async processing stuffs
  # helper function to make async query
  # defp async_query(query) do
  #   caller = self()
  #   spawn(fn ->
  #     send(caller, {:slogged, process_elements(query)})
  #   end)
  # end 

  # # helper function to get result
  # defp get_result(_) do
  #   receive do
  #     {:slogged, result} -> result
  #   end
  # end

  # helper function to process elements
  defp process_elements(d) do
    try do
      any_to_string(d)
    catch
      d ->
        IO.inspect d
        raise SlogInvalidType, message: "Cannot log given value; probably a function! please recheck the values passed."
    end
  end

  # converts all data types to string
  # map checking; matches any map
  defp any_to_string(%{} = m) do
    s = 
      m
      |> Map.to_list()
      |> Stream.map( 
          fn {k, v} -> 
            cond do
              is_atom(k) && k != :__struct__ ->
                "#{k}: #{any_to_string(v)}"
              k == :__struct__ ->
                ""
              true ->
                "#{k} => #{any_to_string(v)}"
            end
          end
        )
      |> Stream.reject(fn e -> String.length(e) == 0 end)
      |> Enum.join(", ")

    # check if it is a struct
    struct_name =
      case Map.fetch(m, :__struct__) do
        {:ok, name} ->
          any_to_string(name)
          |> String.split(".")
          |> List.last()
        _ ->
          ""
      end
      
    # log in map format
    # "%{#{s}}"
    ~s(%#{struct_name}{#{s}})
  end
  # list checking
  defp any_to_string(elem) when is_list(elem) do
    s = 
      elem
      |> Enum.map(&any_to_string/1)
      |> Enum.join(", ")

    "[#{s}]"
  end
  # tuple check
  defp any_to_string(elem) when is_tuple(elem) do
    s = 
      Tuple.to_list(elem)
      |> Enum.map(&any_to_string/1)
      |> Enum.join(", ")
    "{#{s}}"
  end
  # binary check => binary/number/float/boolean
  defp any_to_string(elem) when is_binary(elem) or is_number(elem) or is_float(elem) or is_boolean(elem) do
    to_string(elem)
  end
  # atom check
  defp any_to_string(elem) when is_atom(elem) do
    ":" <> to_string(elem)
  end

  # does not match any of the above guard conditions
  # raise an error
  defp any_to_string(elem) do
    throw(elem)
  end
end

# ref: https://elixir-lang.org/getting-started/try-catch-and-rescue.html#errors
defmodule SlogArgumentError do
  defexception message: "invalid argument to log. Expecting a list ...." 
end

defmodule SlogInvalidType do
  defexception message: "cannot log given element"
end
